File: solutions/python/binary-search/1/binary_search.py
def find(search_list, value):
    def recursiveFind(search_list, value, idx):
        if len(search_list) == 0:
            raise ValueError("value not in array")
        elif len(search_list) == 1:
            if search_list[0] == value:
                return idx
            else:
                raise ValueError("value not in array")
        else:
            middle = search_list[len(search_list) // 2]
            if value == middle:
                return search_list.index(middle) + idx
            elif value < middle:
                return recursiveFind(search_list[:search_list.index(middle)], value, idx)
            else:
                idx += search_list.index(middle) + 1
                return recursiveFind(search_list[search_list.index(middle) + 1:], value, idx)

    return recursiveFind(search_list, value, 0)

File: solutions/fortran/binary-search/1/binary_search.f90
module binary_search
  implicit none

contains

  function find(array, val) result(idx)
    integer, dimension(:), intent(in) :: array
    integer, intent(in) :: val
    integer :: idx
    integer :: low, high, mid

    ! Initialize variables
    low = 1
    high = size(array)
    idx = -1  ! Default to -1 to indicate not found

    ! Perform binary search
    do while (low <= high)
      mid = (low + high) / 2  ! Find the middle index
      if (array(mid) == val) then
        idx = mid  ! Item found at index mid
        return
      elseif (array(mid) < val) then
        low = mid + 1  ! Eliminate the left half
      else
        high = mid - 1  ! Eliminate the right half
      end if
    end do

  end function

end module binary_search
